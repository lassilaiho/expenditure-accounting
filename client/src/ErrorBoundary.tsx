import React from 'react';

import ErrorDisplay from './ErrorDisplay';
import Scaffold from './common/Scaffold';

export interface ErrorBoundaryState {
  currentError: Error | null;
}

export default class ErrorBoundary extends React.Component<
  any,
  ErrorBoundaryState
> {
  public state: ErrorBoundaryState = { currentError: null };

  private onUnhandledRejection = (e: PromiseRejectionEvent) => {
    this.setState({ currentError: toError(e.reason) });
  };

  public componentDidMount() {
    window.addEventListener('unhandledrejection', this.onUnhandledRejection);
  }

  public componentWillUnmount() {
    window.removeEventListener('unhandledrejection', this.onUnhandledRejection);
  }

  public static getDerivedStateFromError(error: unknown) {
    return { currentError: toError(error) };
  }

  public render() {
    if (this.state.currentError === null) {
      return this.props.children;
    }
    return (
      <Scaffold
        title='Error'
        content={<ErrorDisplay error={this.state.currentError} />}
      />
    );
  }
}

function toError(e: unknown) {
  return e instanceof Error ? e : new Error('' + e);
}
