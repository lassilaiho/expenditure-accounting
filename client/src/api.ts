export default class Api {
  public baseUrl: string = '';

  public defaultHeaders: Record<string, string> = {
    'Authorization': ''
  };

  private _sessionToken: string = '';
  public get sessionToken() { return this._sessionToken; }
  public set sessionToken(value: string) {
    this._sessionToken = value;
    this.defaultHeaders['Authorization'] =
      value === '' ? '' : 'Basic ' + btoa(value);
  }

  public constructor(baseUrl: string) {
    this.baseUrl = baseUrl;
  }

  public async postJson(url: string, body: any) {
    return fetch(this.baseUrl + url, {
      mode: 'cors',
      method: 'POST',
      headers: {
        ...this.defaultHeaders,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });
  }
}
