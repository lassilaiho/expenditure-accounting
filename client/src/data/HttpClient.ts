export class ConnectionError extends Error {
  public readonly name = 'ConnectionError';

  public get message() {
    return this.cause.message;
  }

  public constructor(public readonly cause: Error, ...params: any[]) {
    super(...params);

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, ConnectionError);
    }
  }
}

async function apiFetch(input: RequestInfo, init?: RequestInit | undefined) {
  try {
    return await fetch(input, init);
  } catch (e) {
    if (e instanceof Error) {
      throw new ConnectionError(e);
    }
    throw e;
  }
}

export default class HttpClient {
  public baseUrl = '';

  public defaultHeaders: Record<string, string> = {
    Authorization: '',
  };

  private _sessionToken: string | null = null;
  public get sessionToken() {
    return this._sessionToken;
  }
  public set sessionToken(token: string | null) {
    this._sessionToken = token;
    this.defaultHeaders['Authorization'] =
      token === null ? '' : 'Basic ' + btoa(token);
  }

  public constructor(baseUrl: string) {
    this.baseUrl = baseUrl;
  }

  public async get(url: string) {
    return apiFetch(this.baseUrl + url, {
      mode: 'cors',
      headers: { ...this.defaultHeaders },
    });
  }

  public async postJson(url: string, body: any) {
    return apiFetch(this.baseUrl + url, {
      mode: 'cors',
      method: 'POST',
      headers: {
        ...this.defaultHeaders,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });
  }

  public async patchJson(url: string, body: any) {
    return apiFetch(this.baseUrl + url, {
      mode: 'cors',
      method: 'PATCH',
      headers: {
        ...this.defaultHeaders,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });
  }

  public async delete(url: string) {
    return apiFetch(this.baseUrl + url, {
      mode: 'cors',
      method: 'DELETE',
      headers: { ...this.defaultHeaders },
    });
  }
}
