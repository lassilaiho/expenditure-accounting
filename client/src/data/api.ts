export type LoadState =
  | 'not-started'
  | 'loading'
  | 'finished'
  | 'failed';

export class SessionToken {
  public constructor(
    public readonly token: string,
    public readonly expiryTime: Date,
  ) { }

  public static fromJson(json: any): SessionToken | null {
    if (typeof json.token !== 'string'
      || typeof json.expiryTime !== 'string') {
      throw new Error('Invalid json object');
    }
    return new SessionToken(json.token, new Date(json.expiryTime));
  }

  public get isValid() {
    return this.expiryTime.getTime() > Date.now();
  }

  public toJSON(): unknown {
    return {
      token: this.token,
      expiryTime: this.expiryTime.toUTCString(),
    };
  }
}

export default class Api {
  public baseUrl: string = '';

  public defaultHeaders: Record<string, string> = {
    'Authorization': ''
  };

  private _sessionToken: SessionToken | null = null;
  public get sessionToken() { return this._sessionToken; }
  public set sessionToken(token: SessionToken | null) {
    this._sessionToken = token;
    this.defaultHeaders['Authorization'] =
      token === null ? '' : 'Basic ' + btoa(token.token);
  }

  public constructor(baseUrl: string) {
    this.baseUrl = baseUrl;
  }

  public async get(url: string) {
    return fetch(this.baseUrl + url, {
      mode: 'cors',
      headers: { ...this.defaultHeaders },
    });
  }

  public async postJson(url: string, body: any) {
    return fetch(this.baseUrl + url, {
      mode: 'cors',
      method: 'POST',
      headers: {
        ...this.defaultHeaders,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });
  }

  public async patchJson(url: string, body: any) {
    return fetch(this.baseUrl + url, {
      mode: 'cors',
      method: 'PATCH',
      headers: {
        ...this.defaultHeaders,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });
  }

  public async delete(url: string) {
    return fetch(this.baseUrl + url, {
      mode: 'cors',
      method: 'DELETE',
      headers: { ...this.defaultHeaders },
    });
  }
}
