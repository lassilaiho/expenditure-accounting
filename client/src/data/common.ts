import { createAction } from '@reduxjs/toolkit';

import { Product } from './products';
import { Purchase } from './purchases';
import { Tag } from './tags';

export type Collection<T> = {
  byId: { [id: number]: T };
  all: number[];
};

export function emptyCollection<T>(): Collection<T> {
  return { byId: {}, all: [] };
}

export const id = <T>(x: T) => x;

export const clearRemoteData = createAction('clearRemoteData');
export type RemoteDataSet = [Purchase, Product, Tag[]][];
export const setRemoteData = createAction<RemoteDataSet>('setRemoteData');
