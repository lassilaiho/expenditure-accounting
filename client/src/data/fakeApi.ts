import Api from './api';

export type FakeResponse = (
  method: string,
  url: string,
  body?: any,
) => Promise<Response>;

export class FakeApi extends Api {
  public constructor(private fakeResponse: FakeResponse) {
    super('');
  }

  public async get(url: string) {
    return this.fakeResponse('get', url);
  }

  public async postJson(url: string, body: any) {
    return this.fakeResponse('post', url, body);
  }

  public async patchJson(url: string, body: any) {
    return this.fakeResponse('patch', url, body);
  }

  public async delete(url: string) {
    return this.fakeResponse('delete', url);
  }
}

function jsonResponse(body: any) {
  return new Response(JSON.stringify(body), {
    headers: {
      'Content-Type': 'application/json',
    },
    status: 200,
  });
}

const tags = [
  {
    id: 1,
    name: 'tag 1',
  },
  {
    id: 2,
    name: 'tag 2',
  },
  {
    id: 3,
    name: 'tag 3',
  },
];
const products = [
  {
    id: 1,
    name: 'product 1',
  },
  {
    id: 2,
    name: 'product 2',
  },
];
const purchases = [
  {
    id: 1,
    product: products[0],
    date: '2020-03-14',
    quantity: '1.0',
    price: '2.0',
    tags: [tags[0], tags[2]],
  },
  {
    id: 2,
    product: products[0],
    date: '2020-01-14',
    quantity: '10',
    price: '13.2',
    tags: [tags[2]],
  },
  {
    id: 3,
    product: products[1],
    date: '2020-03-14',
    quantity: '0.8',
    price: '0.2',
    tags: [tags[0], tags[1], tags[2]],
  },
];

export const fakeData = { tags, products, purchases };

export const fakeApi = new FakeApi(async (method, url, body) => {
  switch (method + url) {
    case 'get/purchases':
      return jsonResponse({ purchases });
  }
  return new Response(null, { status: 404 });
});
