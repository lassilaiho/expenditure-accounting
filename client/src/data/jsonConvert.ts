import Big from 'big.js';
import moment from 'moment';

export function toNumber(x: any): number {
  if (typeof x === 'number') {
    return x;
  }
  throw new Error('Invalid number ' + x);
}

export function toString(x: any): string {
  if (typeof x === 'string') {
    return x;
  }
  throw new Error('Invalid string ' + x);
}

export function toBig(x: any): Big {
  return new Big(x);
}

export function toObject(x: any): any {
  if (typeof x === 'object') {
    return x;
  }
  throw new Error('Expected json object ' + x);
}

export function toArray<T>(conv: (x: any) => T, arr: any): T[] {
  if (Array.isArray(arr)) {
    return arr.map(conv);
  }
  throw new Error('Expected json array ' + arr);
}

export function toMomentUtc(x: any): moment.Moment {
  const m = moment.utc(x, moment.ISO_8601);
  if (m.isValid()) {
    return m;
  }
  throw new Error('Expected date ' + x);
}
