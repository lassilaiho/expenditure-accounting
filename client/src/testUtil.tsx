import MomentUtil from '@date-io/moment';
import { MuiPickersUtilsProvider } from '@material-ui/pickers';
import { render } from '@testing-library/react';
import React, { useRef } from 'react';
import { MemoryRouter } from 'react-router-dom';

import { fakeApi } from './data/fakeApi';
import { Store, StoreContext } from './data/store';

const TestProviders = (defaultStore?: Store, history?: string[]) => (
  props: any,
) => {
  const store = useRef<Store | null>(null);
  if (store.current === null) {
    if (defaultStore === undefined) {
      store.current = new Store(fakeApi);
    } else {
      store.current = defaultStore;
    }
  }
  return (
    <StoreContext.Provider value={store.current}>
      <MuiPickersUtilsProvider utils={MomentUtil}>
        <MemoryRouter initialEntries={history}>{props.children}</MemoryRouter>
      </MuiPickersUtilsProvider>
    </StoreContext.Provider>
  );
};

export interface CustomRenderOptions {
  store?: Store;
  history?: string[];
}

const customRender = (ui: any, options?: CustomRenderOptions) => {
  let store, history;
  if (options) {
    store = options.store;
    history = options.history;
  }
  return render(ui, { wrapper: TestProviders(store, history) });
};

export * from '@testing-library/react';

export { customRender as render };
