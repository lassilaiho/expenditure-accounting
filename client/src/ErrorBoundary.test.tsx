import React from 'react';

import { ConnectionError } from './data/api';
import ErrorBoundary from './ErrorBoundary';
import { render } from './testUtil';

const RenderErrorComponent: React.FC = () => {
  throw new Error('render error');
};

const ConnectionErrorComponent: React.FC = () => {
  throw new ConnectionError(new Error('no connection'));
};

test('renders correctly when no error occurs', () => {
  const { asFragment } = render(
    <ErrorBoundary>
      <div>test</div>
    </ErrorBoundary>,
  );
  expect(asFragment()).toMatchSnapshot();
});

test('handles an error during rendering correctly', () => {
  const spy = jest.spyOn(console, 'error');
  spy.mockImplementation(() => undefined);

  const { asFragment } = render(
    <ErrorBoundary>
      <RenderErrorComponent />
    </ErrorBoundary>,
  );
  expect(asFragment()).toMatchSnapshot();

  spy.mockRestore();
});

test('handles a connection error correctly', () => {
  const spy = jest.spyOn(console, 'error');
  spy.mockImplementation(() => undefined);

  const { asFragment } = render(
    <ErrorBoundary>
      <ConnectionErrorComponent />
    </ErrorBoundary>,
  );
  expect(asFragment()).toMatchSnapshot();

  spy.mockRestore();
});
