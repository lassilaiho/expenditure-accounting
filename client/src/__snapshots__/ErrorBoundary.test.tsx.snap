// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles a connection error correctly 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-7"
  >
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <h6
          class="MuiTypography-root makeStyles-title-9 MuiTypography-h6"
        >
          Error
        </h6>
      </div>
    </header>
    <div
      class="makeStyles-contentRoot-8"
    >
      <div
        class="MuiContainer-root MuiContainer-fixed MuiContainer-maxWidthLg"
        style="width: 100%; height: 100%;"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            class="MuiBox-root MuiBox-root-10"
          >
            <h4
              class="MuiTypography-root MuiTypography-h4"
            >
              Network connection is unavailable.
            </h4>
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph"
            >
              You can try reloading the page when you are back online.
            </p>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Reload page
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <br />
            <br />
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Show details
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`handles an error during rendering correctly 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-2"
  >
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <h6
          class="MuiTypography-root makeStyles-title-4 MuiTypography-h6"
        >
          Error
        </h6>
      </div>
    </header>
    <div
      class="makeStyles-contentRoot-3"
    >
      <div
        class="MuiContainer-root MuiContainer-fixed MuiContainer-maxWidthLg"
        style="width: 100%; height: 100%;"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            class="MuiBox-root MuiBox-root-5"
          >
            <h4
              class="MuiTypography-root MuiTypography-h4"
            >
              Something went wrong.
            </h4>
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph"
            >
              You can try reloading the page.
            </p>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Reload page
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <br />
            <br />
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Show details
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders correctly when no error occurs 1`] = `
<DocumentFragment>
  <div>
    test
  </div>
</DocumentFragment>
`;
